internal/powerstore/powerstore.go

```diff
@@ -0,0 +1,350 @@
+	for baseEpoch < targetEpoch {
+		baseEpoch++
+		diff, err := ps.get(ctx, baseEpoch)
+		if err != nil {
+			return nil, err
+		}
+		diffs = append(diffs, diff)
+	}
+	return certs.ApplyPowerTableDiffs(basePt, diffs...)
+}
+
+func (ps *Store) basePowerTable(ctx context.Context) (int64, gpbft.PowerEntries, error) {
+	baseEpoch := ps.manifest.BootstrapEpoch - ps.manifest.ECFinality
+	baseInstance := ps.manifest.InitialInstance
+	if lastCert := ps.cs.Latest(); lastCert != nil {
+		baseInstance = lastCert.GPBFTInstance + 1
+		if lastCert.GPBFTInstance > ps.manifest.InitialInstance+ps.manifest.CommitteeLookback {
+			baseCert, err := ps.cs.Get(ctx, lastCert.GPBFTInstance-ps.manifest.CommitteeLookback)
+			if err != nil {
+				return 0, nil, err
+			}
+			baseEpoch = baseCert.ECChain.Head().Epoch
+		}
+	}
+
+	basePt, err := ps.cs.GetPowerTable(ctx, baseInstance)
+	if err != nil {
+		return 0, nil, fmt.Errorf("failed to load base power table: %w", err)
+	}
+	return baseEpoch, basePt, nil
+}
+
+func (ps *Store) mostRecentPowerTable(ctx context.Context) (int64, gpbft.PowerEntries, error) {
+	baseEpoch, basePt, err := ps.basePowerTable(ctx)
+	if err != nil {
+		return 0, nil, err
+	}
+
+	var diffs []certs.PowerTableDiff
+	for {
+		diff, err := ps.get(ctx, baseEpoch+1)
+		if err != nil {
+			if !errors.Is(err, datastore.ErrNotFound) {
+				log.Errorw("failed to lookup power table delta", "epoch", baseEpoch+1, "error", err)
+			}
+			break
+		}
+		diffs = append(diffs, diff)
+		baseEpoch++
+	}
+	basePt, err = certs.ApplyPowerTableDiffs(basePt, diffs...)
+	if err != nil {
+		return 0, nil, err
+	}
+	return baseEpoch, basePt, nil
+}
+
+func (ps *Store) put(ctx context.Context, epoch int64, diff certs.PowerTableDiff) error {
+	var buf bytes.Buffer
+	if err := diff.MarshalCBOR(&buf); err != nil {
+		return err
+	}
+	if err := ps.ds.Put(ctx, ps.dsKeyForDiff(epoch), buf.Bytes()); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (ps *Store) f3HeadEpoch() int64 {
+	latestCert := ps.cs.Latest()
+	if latestCert == nil {
+		return ps.manifest.BootstrapEpoch - ps.manifest.ECFinality
+	}
+	return latestCert.ECChain.Head().Epoch
+}
+
+func (ps *Store) deleteAll(ctx context.Context) {
+	res, err := ps.ds.Query(ctx, query.Query{
+		Prefix: diffPrefix,
+		Orders: []query.Order{query.OrderByKey{}},
+	})
+	if err != nil {
+		log.Errorw("failed to query for power-table diffs to delete", "error", err)
+		return
+	}
+	defer res.Close()
+	for r, ok := res.NextSync(); ok; r, ok = res.NextSync() {
+		if r.Error != nil {
+			log.Errorw("failed to query for power-table diffs to delete", "error", r.Error)
+			continue
+		}
+
+		if err := ps.ds.Delete(ctx, datastore.NewKey(r.Key)); err != nil {
+			log.Errorw("failed to delete power-table diffs to delete", "error", err)
+			return
+		}
+	}
+}
+
+func (ps *Store) run(ctx context.Context) error {
+	ticker := ps.clock.Ticker(ps.manifest.ECPeriod * time.Duration(max(ps.manifest.ECFinality/4, 1)))
```

Comment:
This might have weird interactions when a node is catching up. I suggest we do it more frequently; there is no harm in it, really.

Reply:
Fair enough. I can also make it faster when we're behind.

Reply:
Eh, I just made it every 2-ec.